File: ios/Sources/MeasureSDK/Swift/Exporter/BatchCreator.swift
//
//  BatchCreator.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 17/10/24.
//

import Foundation

protocol BatchCreator {
    func create(sessionId: String?, completion: @escaping ((BatchCreationResult?) -> Void))
}

typealias BatchCreationResult = (batchId: String, eventIds: [String], spanIds: [String])

final class BaseBatchCreator: BatchCreator {
    private let logger: Logger
    private let idProvider: IdProvider
    private let configProvider: ConfigProvider
    private let timeProvider: TimeProvider
    private let eventStore: EventStore
    private let batchStore: BatchStore
    private let spanStore: SpanStore

    init(logger: Logger, idProvider: IdProvider, configProvider: ConfigProvider, timeProvider: TimeProvider, eventStore: EventStore, batchStore: BatchStore, spanStore: SpanStore) {
        self.logger = logger
        self.idProvider = idProvider
        self.configProvider = configProvider
        self.timeProvider = timeProvider
        self.eventStore = eventStore
        self.batchStore = batchStore
        self.spanStore = spanStore
    }

    func create(sessionId: String? = nil, completion: @escaping (BatchCreationResult?) -> Void) {
        eventStore.getUnBatchedEventsWithAttachmentSize(
            eventCount: configProvider.maxEventsInBatch,
            ascending: true,
            sessionId: sessionId
        ) { [weak self] eventToAttachmentSizeMap in
            guard let self else { return }

            let eventIds = filterEventsForMaxAttachmentSize(eventToAttachmentSizeMap)

            spanStore.getUnBatchedSpans(
                spanCount: configProvider.maxEventsInBatch,
                ascending: true
            ) { [weak self] spanIds in
                guard let self else { return }

                if eventIds.isEmpty && spanIds.isEmpty {
                    logger.log(level: .debug, message: "No events or spans to batch.", error: nil, data: nil)
                    completion(nil)
                    return
                }

                let batchId = idProvider.uuid()
                let batch = BatchEntity(
                    batchId: batchId,
                    eventIds: eventIds,
                    spanIds: spanIds,
                    createdAt: timeProvider.now()
                )

                batchStore.insertBatch(batch) { [weak self] isBatchInsertionSuccessful in
                    guard let self else { return }

                    if !isBatchInsertionSuccessful {
                        logger.log(level: .error, message: "Failed to insert batched event IDs", error: nil, data: nil)
                        completion(nil)
                        return
                    }

                    // Fire-and-forget updates
                    eventStore.updateBatchId(batchId, for: eventIds)
                    spanStore.updateBatchId(batchId, for: spanIds)

                    let result = BatchCreationResult(
                        batchId: batchId,
                        eventIds: eventIds,
                        spanIds: spanIds
                    )
                    completion(result)
                }
            }
        }
    }

    /// Filters events to ensure the total size of attachments does not exceed the maximum limit.
    private func filterEventsForMaxAttachmentSize(_ eventToAttachmentSizeMap: [String: Int64]) -> [String] {
        var totalSize: Int64 = 0
        return eventToAttachmentSizeMap.prefix {
            totalSize += $0.value
            return totalSize <= configProvider.maxAttachmentSizeInEventsBatchInBytes
        }.map { $0.key }
    }
}

File: android/measure/src/main/java/sh/measure/android/exporter/BatchCreator.kt
package sh.measure.android.exporter

import sh.measure.android.config.ConfigProvider
import sh.measure.android.logger.LogLevel
import sh.measure.android.logger.Logger
import sh.measure.android.storage.BatchEntity
import sh.measure.android.storage.Database
import sh.measure.android.utils.IdProvider
import sh.measure.android.utils.TimeProvider

/**
 * Creates a batch of events to be exported.
 */
internal interface BatchCreator {
    /**
     * Attempts to create a new batch of events to export.
     *
     * @param sessionId The session ID to filter events by if provided.
     * @return [Batch] if a batch was created, otherwise null.
     */
    fun create(sessionId: String? = null): Batch?
}

/**
 * Result of a batch creation operation. Contains the mapping of batch ID to all the event IDs part
 * of the batch.
 */
internal data class Batch(
    val batchId: String,
    val eventIds: List<String>,
    val spanIds: List<String>,
)

/**
 * Implementation of [BatchCreator] which creates a new batch from all available un-batched events
 * in database, keeping all configuration limits in mind. It also synchronizes batch creation by
 * ensuring only one batch can be created at a time.
 */
internal class BatchCreatorImpl(
    private val logger: Logger,
    private val idProvider: IdProvider,
    private val database: Database,
    private val configProvider: ConfigProvider,
    private val timeProvider: TimeProvider,
) : BatchCreator {
    private val batchCreationLock = Any()

    override fun create(sessionId: String?): Batch? {
        synchronized(batchCreationLock) {
            val eventToAttachmentSizeMap =
                database.getUnBatchedEventsWithAttachmentSize(
                    configProvider.maxEventsInBatch,
                    sessionId = sessionId,
                    eventTypeExportAllowList = configProvider.eventTypeExportAllowList,
                )
            if (eventToAttachmentSizeMap.isEmpty()) {
                return null
            }

            val eventIds = filterEventsForMaxAttachmentSize(eventToAttachmentSizeMap)
            val spanIds = database.getUnBatchedSpans(configProvider.maxEventsInBatch)

            if (spanIds.isEmpty() && eventIds.isEmpty()) {
                logger.log(LogLevel.Debug, "No events or spans to batch")
            }

            val batchId = idProvider.uuid()
            val batchInsertionResult = database.insertBatch(
                BatchEntity(
                    batchId = batchId,
                    eventIds = eventIds,
                    spanIds = spanIds,
                    createdAt = timeProvider.now(),
                ),
            )
            if (!batchInsertionResult) {
                logger.log(LogLevel.Debug, "Failed to insert batch")
                return null
            }
            return Batch(
                batchId = batchId,
                eventIds = eventIds,
                spanIds = spanIds,
            )
        }
    }

    private fun filterEventsForMaxAttachmentSize(eventToAttachmentSizeMap: LinkedHashMap<String, Long>): List<String> {
        var totalSize = 0L
        return eventToAttachmentSizeMap.asSequence().takeWhile { (_, size) ->
            totalSize += size
            totalSize <= configProvider.maxAttachmentSizeInEventsBatchInBytes
        }.map { (key, _) -> key }.toList()
    }
}
